clang/lib/Sema/SemaStmtAttr.cpp

```diff
@@ -554,11 +554,52 @@ void CheckForIncompatibleUnrollHintAttributes(
     SourceLocation Loc = Range.getBegin();
     S.Diag(Loc, diag::err_loop_unroll_compatibility)
         << PragmaUnroll->getDiagnosticName(Policy)
-        << AttrUnroll->getDiagnosticName();
+        << AttrUnroll->getDiagnosticName(Policy);
   }
 }
 
-template <typename LoopUnrollAttrT>
+static bool HandleLoopUnrollAttrExpr(Sema &S, Expr *E,
+                                     const AttributeCommonInfo &A,
+                                     unsigned *UnrollFactor = nullptr) {
+  if (E && !E->isInstantiationDependent()) {
+    llvm::APSInt ArgVal(32);
+
+    if (!E->isIntegerConstantExpr(ArgVal, S.Context)) {
+      S.Diag(E->getExprLoc(), diag::err_attribute_argument_type)
+          << A.getAttrName() << AANT_ArgumentIntegerConstant
+          << E->getSourceRange();
+      return false;
```

Comment:
The bool return type seems weird.  Typically these return the attribute themselves or null.  Additionally, in clang, 'true' is usually the 'failure' value.

Reply:
Is it better to rename function to `IsValidLoopUnrollAttrExpr` or to change return statements from false to true and backwards?

Reply:
Is there a good reason it returns bool rather than creating the attribute like the rest of the functions like this do?  If so, `CheckLoopUnrollAttrExpr` is likely a better name(more consistent) with 'true' meaning a failure (it is the common error/non-error flag in clang, also allows "return S.Diag(...").

Reply:
The reason is that OpenCL attribute accepts unsigned value as an argument, and SYCL attribute accepts an expression - this make impossible to unify attributes creation. The goal was to keep diagnostics unified for both attributes